File: 589. N-ary Tree Preorder Traversal/recursive.py
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        res = []
        self.traversal(res, root)
        return res
        
    def traversal(self, res: List[int], root: 'Node'):
        if root is None:
            return
            
        res.append(root.val)
        
        for child in root.children:
            self.traversal(res, child)

File: 589. N-ary Tree Preorder Traversal/recursive.cpp
class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> res;
        traverse(root, res);
        return res;
    }
    
    void traverse(Node* root, vector<int>& res) {
        if (root == nullptr) return;
        
        res.push_back(root->val);
        
        for (auto node: root->children)
            traverse(node, res);
    }
};